web/profiles/custom/os2loop/modules/os2loop_oembed/src/TwigExtension/TwigExtension.php

```diff
@@ -0,0 +1,254 @@
+use Drupal\os2loop_settings\Settings;
+
+/**
+ * Custom twig functions.
+ */
+class TwigExtension extends AbstractExtension {
+  use StringTranslationTrait;
+
+  /**
+   * Guzzle\Client instance.
+   *
+   * @var \GuzzleHttp\ClientInterface
+   */
+  protected $httpClient;
+
+  /**
+   * The url resolver service.
+   *
+   * @var \Drupal\media\OEmbed\UrlResolverInterface
+   */
+  protected $urlResolver;
+
+  /**
+   * The Messenger service.
+   *
+   * @var \Drupal\Core\Messenger\MessengerInterface
+   */
+  protected $messenger;
+
+  /**
+   * The serializer service.
+   *
+   * @var \Symfony\Component\Serializer\SerializerInterface
+   */
+  protected $serializer;
+
+  /**
+   * Drupal\Core\Render\RendererInterface definition.
+   *
+   * @var \Drupal\Core\Render\RendererInterface
+   */
+  protected $renderer;
+
+  /**
+   * The settings.
+   *
+   * @var \Drupal\os2loop_settings\Settings
+   */
+  protected $settings;
+
+  /**
+   * TwigExtension constructor.
+   *
+   * @param \GuzzleHttp\ClientInterface $http_client
+   *   The http client.
+   * @param \Drupal\media\OEmbed\UrlResolverInterface $urlResolver
+   *   The Url resolver service.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   The messenger service.
+   * @param \Symfony\Component\Serializer\SerializerInterface $serializer
+   *   The serializer service.
+   * @param \Drupal\Core\Render\RendererInterface $renderer
+   *   Drupals renderer service.
+   * @param \Drupal\os2loop_settings\Settings $settings
+   *   Loop settings.
+   */
+  public function __construct(ClientInterface $http_client, UrlResolverInterface $urlResolver, MessengerInterface $messenger, SerializerInterface $serializer, RendererInterface $renderer, Settings $settings) {
+    $this->httpClient = $http_client;
+    $this->urlResolver = $urlResolver;
+    $this->messenger = $messenger;
+    $this->serializer = $serializer;
+    $this->renderer = $renderer;
+    $this->settings = $settings;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFilters(): array {
+    return [
+      new TwigFilter('create_video', [$this, 'createVideo']),
+    ];
+  }
+
+  /**
+   * Render a video from an embed url or iframe.
+   *
+   * @param array $text
+   *   The input text that the filter was applied to.
+   *
+   * @return \Drupal\Component\Render\MarkupInterface
+   *   The rendered html.
+   *
+   * @throws \GuzzleHttp\Exception\GuzzleException
+   *   An exception if guzzle fails.
+   * @throws \Exception
+   *   The genral exception.
+   */
+  public function createVideo(array $text): MarkupInterface {
+    // Set string depending on field type.
+    $string = isset($text['#plain_text']) ? $text['#plain_text'] : $text['#context']['value'];
+    if (!empty($this->findIframe($string))) {
+      $string = $this->handleIframe($string);
+    }
+    $videoArray = $this->createVideoFromUrl($string);
+    $videoArray = $this->applyCookieConsent($videoArray);
```

Comment:
If I understand oembed correctly, then it basically a way to convert a url to embed markup (typically an `iframe`), and I this we should reflect this more clearly here by writing something along the lines of
```
$url = $this->getUrl($text); // Returns a string
$markup = $this->getMarkupFromUrl($url); // Returns a string
return $this->applyCookieConsent($markup); // Returns a string 
```

Passing arrays around seem cumbersome and I don't think we need more than strings.

Reply:
I think this will result in a greater rewrite that can not be contained within the estimate.